File: Lab2.3/lab2_3.cpp
#include <vector>
#include <fstream>
#include <iostream>
#include <cmath>
using namespace std;

double y(double x) {
	return (sqrt(x) + cos(x));
}

double Integrate(double a, double b, int count, double (*f) (double)) {
	double sum = 0;
	double h = (b - a) / count;
	for (int i = 0; i < count; i++) {
		sum += (h)*f(a + h * i + h / 2);
	}
	return sum;
}
void AbsErrPlot() /*abs err of num of nodes*/ {
	ofstream out("approx_count.csv");
	const double a = 0.1;
	const double b = 1;
	const double accurate_value = 1.387222383759946;
	int count = 2;
	double cur = 0;
	double eps = 1e-1;
	for (eps; eps > 1e-12; eps /= 10) {
		count = 2;
		do {
			cur = Integrate(a, b, count, y);
			out.precision(16);
			count *= 2;
		} while (abs(accurate_value - cur) > eps);
		double theormax = (b - a) / 24 * (8.901) * pow(((b - a) / count), 2);
		double theormin = (b - a) / 24 * (0.79) * pow(((b - a) / count), 2);
		out << abs(accurate_value - cur) << ';' << count << ';' << theormax << ';' << theormin << ';' << endl;
	}
	return;
}

void NodesEpsPlot() {
	ofstream out("nodes_eps.csv");
	const double a = 0.1;
	const double b = 1;
	for (double eps = 0.01; eps >= 1e-12; eps /= 10) {
		int count = 2;
		double prev = 0;
		double cur = Integrate(a, b, count, y);
		do {
			prev = cur;
			count *= 2;
			cur = Integrate(a, b, count, y);
		} while (fabs(cur - prev) > eps);
		out.precision(16);
		out << fabs(cur - prev) << ';' << count << ';' << eps << ';' << endl;
	}
}

int main() {
	
	AbsErrPlot();
	NodesEpsPlot();
	return 0;
}

File: Lab2.3/lab2_3.m
f = @(x) sqrt(x) + cos(x);
absErrPlot = importdata ('approx_count.csv');
nodesEpsPlot = importdata ('nodes_eps.csv');
accurate = integral(f, 0.1, 1);
for i = 1:12
    eps(i) = 1.* 10.^(-i);
end
% for i = 1:11
%     abs_err(i) = abs(accurate - absErrPlot(i,1));
% end
% for i = 1:11
%     abs_err_thr(i) = abs(accurate - nodesEpsPlot(i,1))
% end

figure;
loglog(eps(1:11),nodesEpsPlot(:,1));
hold on
grid on
loglog(eps,eps);
xlabel('eps');
ylabel('Runge error');
title('Runge error (given epsilon)')
legend({'Runge error','y = x (eps)'},'Location','southeast')

figure;
loglog(nodesEpsPlot(:,3), nodesEpsPlot(:,2));
hold on
grid on
xlabel('eps');
ylabel('num of intervals');
legend({'N(\epsilon)'},'Location','southwest')
title('Num of intervals (epsilon)')

figure;
loglog(eps,absErrPlot(:,1));
hold on
loglog(eps,eps)
loglog(eps, absErrPlot(:,3));
loglog(eps, absErrPlot(:,4));
legend({'y = x (eps)', 'actual','theormax','theormin'},'Location','northwest')
xlabel('eps');
ylabel('errors (min, max, actual (|acc - approx|)');
title('error(\epsilon)')